crates/sui-surfer/src/default_surf_strategy.rs

```diff
@@ -0,0 +1,200 @@
+                            break;
+                        }
+                    }
+                }
+                Type::MutableReference(ty) => {
+                    match Self::choose_object_call_arg(
+                        state,
+                        InputObjectPassKind::MutRef,
+                        *ty,
+                        &mut chosen_owned_objects,
+                    )
+                    .await
+                    {
+                        Some(arg) => arg,
+                        None => {
+                            failed = true;
+                            break;
+                        }
+                    }
+                }
+                Type::U256 | Type::Signer | Type::Vector(_) | Type::TypeParameter(_) => {
+                    failed = true;
+                    break;
+                }
+            };
+            args.push(arg);
+        }
+        if failed {
+            for (struct_tag, obj_ref) in chosen_owned_objects {
+                state
+                    .owned_objects
+                    .get_mut(&struct_tag)
+                    .unwrap()
+                    .insert(obj_ref);
+            }
+            None
+        } else {
+            Some(args)
+        }
+    }
+
+    async fn choose_object_call_arg(
+        state: &mut SurferState,
+        kind: InputObjectPassKind,
+        arg_type: Type,
+        chosen_owned_objects: &mut Vec<(StructTag, ObjectRef)>,
+    ) -> Option<CallArg> {
+        let type_tag = match arg_type {
+            Type::Struct {
+                address,
+                module,
+                name,
+                type_arguments,
+            } => StructTag {
+                address,
+                module,
+                name,
+                type_params: type_arguments
+                    .into_iter()
+                    .map(|t| t.into_type_tag().unwrap())
+                    .collect(),
+            },
+            _ => {
+                return None;
+            }
+        };
+        let owned = state.matching_owned_objects_count(&type_tag);
+        let immutable = state.matching_immutable_objects_count(&type_tag).await;
+        let shared = state.matching_shared_objects_count(&type_tag).await;
+
+        let total_matching_count = match kind {
+            InputObjectPassKind::Value => owned,
+            InputObjectPassKind::ByRef => owned + immutable + shared,
+            InputObjectPassKind::MutRef => owned + shared,
+        };
+        if total_matching_count == 0 {
+            return None;
+        }
+        let mut n = state.rng.gen_range(0..total_matching_count);
+        if n < owned {
+            let obj_ref = state.choose_nth_owned_object(&type_tag, n);
+            chosen_owned_objects.push((type_tag, obj_ref));
+            return Some(CallArg::Object(ObjectArg::ImmOrOwnedObject(obj_ref)));
+        }
+        n -= owned;
+        if matches!(kind, InputObjectPassKind::ByRef) {
```

Comment:
I think if we swap the order and deal with shared objects before immutable objects, then there's no need to do this check here. Since if we get to that stage, `n` must be greater than `owned + shared` so the pass kind must be `ByRef`.